keps/sig-api-machinery/4222-cbor-serializer/README.md

```diff
@@ -0,0 +1,1274 @@
+interpretation of such an object is unambiguous.
+
+A map containing duplicate keys is well-formed but invalid according to the CBOR
+specification. Decoding a map containing duplicate keys will produce a decode
+error.
+
+<<[UNRESOLVED]>>
+How should the decoder handle unrecognized fields, that is, encountering a map
+key that does not correspond to the name of a struct field's `json` tag name?
+This may be expected in cases where the client is newer than the server, or
+where an object containing an unrecognized field was transcoded from YAML or
+JSON to CBOR. A strict decode error (as in JSON) may be appropriate in this case.
+
+This is not a concern for the custom resource path, which uses a schema-aware
+decoder wrapper to detect unknown fields.
+<<[/UNRESOLVED]>>
+
+Note that clients (e.g. kubectl) may choose to decode an object from a JSON or
+YAML text representation containing duplicate keys, then encode to CBOR to
+populate the body of an API request. Since the text-encoded content (potentially
+containing duplicates) is not literally transcoded to CBOR, this use case is
+supported. Depending on the strictness mode, duplicate keys would either be
+removed or produce an error at decode time.
+
+### Encoding Determinism
+
+It is possible for a single object to be encodable as multiple distinct but
+valid and semantically-equivalent CBOR byte strings. The CBOR specification does
+not require encoder implementations to produce deterministic output, although it
+does include recommendations for implementing deterministic encoding.
+
+The etcd3 storage implementation of GuaranteedUpdate relies on deterministic
+encoding to skip writes if the stored bytes would not change. The existing JSON
+and Protobuf encoders produce deterministic output.
+
+Other potential use cases for deterministic encoding of response bodies might
+include:
+
+- caching
+  - The existing WithCacheControl filter sets the response header
+    “Cache-Control: no-cache, private” to prevent shared caches from storing
+    responses (since requests are subject to authn/authz), and to prevent
+    responses in non-shared caches from being reused without
+    validation. Deterministic encoding could allow an API server to generate
+    strong ETags by hashing the encoded form of the resource.
+  - Even for the existing data formats, there should be no caching proxies
+    storing API responses.
+- diffing
+  - The human-readable text formats (JSON and YAML) are not changing under this
+    proposal.
+
+Encode benchmarks for the two evaluated Go CBOR libraries show a 2.4x speedup
+and a 1.8x speedup by disabling map key sorting. According to the spec, “the
+CBOR data model for maps does not allow ascribing semantics to the order of the
+key/value pairs in the map representation.”  And since the CBOR decoder will
+reject maps containing duplicate keys, a CBOR map represents exactly the same
+set of key-value pairs regardless of the order they are encoded.
```

Comment:
I would forget about that. The same is true for json, yaml, etc. There are speed-up to serializing json in a non-deterministic way, we've just decided to skip these.

Reply:
I agree we can't change this behavior in an existing format. If the decision is against a nondeterministic encoding mode, that is probably irreversible as with JSON and YAML. For a new format, and given the motivation, I think it's fair to consider the potential speedup and whether or not the risks can be mitigated.